core/frontend/src/render/webgl/Texture.ts

```diff
@@ -480,6 +511,88 @@ export class Texture2DHandle extends TextureHandle {
   }
 }
 
+/** @internal */
+export interface ExternalTextureRequest {
+  handle: Texture2DHandle;
+  name: string;
+  imodel: IModelConnection;
+  type: RenderTexture.Type;
+  format: ImageSourceFormat;
+  onLoaded?: (req: ExternalTextureRequest) => void;
+}
+
+/** @internal */
+export class ExternalTextureLoader { /* currently exported for tests only */
+  public static readonly instance = new ExternalTextureLoader(10);
+  private readonly _maxActiveRequests: number;
+  private _activeRequests: Array<ExternalTextureRequest> = [];
+  private _pendingRequests: Array<ExternalTextureRequest> = [];
+
+  public get numActiveRequests() { return this._activeRequests.length; }
+  public get numPendingRequests() { return this._pendingRequests.length; }
+  public get maxActiveRequests() { return this._maxActiveRequests; }
+
+  private constructor(maxActiveRequests: number) {
+    this._maxActiveRequests = maxActiveRequests;
+  }
+
+  private async _nextRequest(prevReq: ExternalTextureRequest) {
+    this._activeRequests.splice(this._activeRequests.indexOf(prevReq), 1);
+    if (this._activeRequests.length < this._maxActiveRequests && this._pendingRequests.length > 0) {
+      const req = this._pendingRequests.pop()!;
+      await this._activateRequest(req);
+    }
+  }
+
+  private async _activateRequest(req: ExternalTextureRequest) {
+    if (req.imodel.isClosed)
+      return;
+
+    this._activeRequests.push(req);
+
+    try {
+      if (!req.imodel.isClosed) {
+        const texBytes = await req.imodel.getTextureImage({ name: req.name });
+        if (undefined !== texBytes) {
+          const imageSource = new ImageSource(texBytes, req.format);
+          const image = await imageElementFromImageSource(imageSource);
+          if (!req.imodel.isClosed) {
+            req.handle.reload(Texture2DCreateParams.createForImage(image, ImageSourceFormat.Png === req.format, req.type));
+            IModelApp.tileAdmin.invalidateAllScenes();
+            if (undefined !== req.onLoaded)
+              req.onLoaded(req);
+          }
+        }
+      }
+    } catch (_e) { }
+
+    await this._nextRequest(req);

```

Comment:
/s/await/return

Reply:
Just that one line - right?  the final called to _nextRequest().  The awaits within the try-block need to remain.

Reply:
Yes, just this last line. BTW you'll need to run extract-api on imodeljs-frontend again.

Reply:
pushed the await change.  extract-api isn't picking up any changes for me with the current set of changes...

Reply:
Hmm, I expected api changes since you changed the return type of a couple of functions by marking them `async`. Maybe by not explicitly declaring return type it didn't pick up on it.

Reply:
Yeah, I kind of expected it too.